src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php

```diff
@@ -147,76 +168,102 @@ protected function isExcluded($url)
+            $tryPublicUrls[rtrim($publicUrl, '/')] = $this->slashSuffixHandling;
+            if ($queryString !== '') {
+                $tryPublicUrls[rtrim($publicUrl, '/') . $queryString] = $this->slashSuffixHandling;
+            }
+        }
 
-                $publicUrl = substr($publicUrl, 0, strpos($publicUrl, '?'));
-                if ($url = $this->aliasing->hasInternalAlias($publicUrl, true, UrlAlias::REWRITE)) {
-                    $this->rewriteRequest($event, $url->getInternalUrl());
+        $qb = $this->aliasing->getRepository()->createQueryBuilder('u');
+        $qb->where($qb->expr()->in('u.public_url', array_keys($tryPublicUrls)))
+            ->indexBy('u', 'u.public_url');
+        /** @var UrlAlias[] $urlAliases */
+        $urlAliases = $qb->getQuery()->getResult();
 
-                    return;
-                }
+        if (count($urlAliases) === 0) {
+            return;
+        }
+
+        foreach ($tryPublicUrls as $tryPublicUrl => $handlingMode) {
+            if (!array_key_exists($tryPublicUrl, $urlAliases)) {
+                continue;
+            }
+
+            $urlAlias = $urlAliases[$tryPublicUrl];
+
+            switch ($handlingMode) {
+                case static::SLASH_SUFFIX_REDIRECT_TEMP:
+                    $url = $urlAlias->getMode() === UrlAlias::ALIAS ? $urlAlias->getInternalUrl() : $urlAlias->getPublicUrl(); // Same mode? Use internal URL directly, don't go redirecting twice...
+                    $event->setResponse(new RedirectResponse($url, Response::HTTP_FOUND));
+                    break;
+
+                case static::SLASH_SUFFIX_REDIRECT_PERM:
+                    $url = $urlAlias->getMode() === UrlAlias::MOVE ? $urlAlias->getInternalUrl() : $urlAlias->getPublicUrl(); // Same mode? Use internal URL directly, don't go redirecting twice...
+                    $event->setResponse(new RedirectResponse($url, Response::HTTP_MOVED_PERMANENTLY));
+                    break;
+
+                case static::SLASH_SUFFIX_ACCEPT:
+                    // Continue as if the correct URL (without '/' suffix) was requested. Could result in duplicate content disqualifications
+                default: // $handlingMode === null
+                    switch ($urlAlias->getMode()) {
+                        case UrlAlias::REWRITE:
+                            $this->rewriteRequest($event, $urlAlias->getInternalUrl());
+                            break;
+                        case UrlAlias::MOVE:
+                        case UrlAlias::ALIAS:
+                            $event->setResponse(new RedirectResponse($urlAlias->getInternalUrl(), $urlAlias->getMode()));
+                            break;
+                        default:
+                            throw new \UnexpectedValueException(sprintf("Invalid mode %s for UrlAlias %s.", $urlAlias->getMode(), json_encode($urlAlias)));
+                    }
+                break;
```

Comment:
Ik denk dat het heel verstandig is om unit tests te hebben om zeker te weten dat de oude functionaliteit gewaarborgd is en de nieuwe vast staat...

Reply:
Er zijn wel unit tests. Die slaan voor de helft nergens op zag ik zo 1,2,3

Reply:
@boudewijn-zicht je geeft aan dat omdat het een kritiek stukje code is dat je eigenlijk wel wilt dat er tests zijn. Ik zie, nogmaals, vanwege de kwaliteit van de huidige tests hier op dit moment geen meerwaarde in.

Als ik phpunit nu draai, dan krijg ik onderstaande output (+ nog 20 PHP errors). Er is dus werk aan de winkel, maar niet in deze PR. Wel kun je zien dat de tests die er zijn, niet falen op mijn wijzigingen.

Huidige test output:

```
PHPUnit 4.8.36 by Sebastian Bergmann and contributors.

..........................................EEE..EEEEE...........  63 / 143 ( 44%)
............................................................... 126 / 143 ( 88%)
.................

Time: 1.37 seconds, Memory: 18.00MB

There were 8 errors:

1) ZichtTest\Bundle\UrlBundle\Aliasing\ListenerTest::testOnKernelRequestHandlesMasterRequest
Error: Call to a member function createQueryBuilder() on null

/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php:220
/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/tests/ZichtTest/Bundle/UrlBundle/Aliasing/ListenerTest.php:53

2) ZichtTest\Bundle\UrlBundle\Aliasing\ListenerTest::testRequestIsRoutedWithInternalUrl
Error: Call to a member function createQueryBuilder() on null

/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php:220
/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/tests/ZichtTest/Bundle/UrlBundle/Aliasing/ListenerTest.php:78

3) ZichtTest\Bundle\UrlBundle\Aliasing\ListenerTest::testParameterParsing
Error: Call to a member function createQueryBuilder() on null

/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php:220
/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/tests/ZichtTest/Bundle/UrlBundle/Aliasing/ListenerTest.php:100

4) ZichtTest\Bundle\UrlBundle\Aliasing\ListenerTest::testUrlExclusion with data set #1 (true, '/\bqux\b/', 'foo bar bat baz')
Error: Call to a member function createQueryBuilder() on null

/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php:220
/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/tests/ZichtTest/Bundle/UrlBundle/Aliasing/ListenerTest.php:148

5) ZichtTest\Bundle\UrlBundle\Aliasing\ListenerTest::testRedirectModes with data set #0 (301)
Error: Call to a member function createQueryBuilder() on null

/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php:220
/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/tests/ZichtTest/Bundle/UrlBundle/Aliasing/ListenerTest.php:185

6) ZichtTest\Bundle\UrlBundle\Aliasing\ListenerTest::testRedirectModes with data set #1 (302)
Error: Call to a member function createQueryBuilder() on null

/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php:220
/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/tests/ZichtTest/Bundle/UrlBundle/Aliasing/ListenerTest.php:185

7) ZichtTest\Bundle\UrlBundle\Aliasing\ListenerTest::testRedirectModes with data set #2 (-1, true)
Error: Call to a member function createQueryBuilder() on null

/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php:220
/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/tests/ZichtTest/Bundle/UrlBundle/Aliasing/ListenerTest.php:177

8) ZichtTest\Bundle\UrlBundle\Aliasing\ListenerTest::testParameterParsingUtf8
Error: Call to a member function createQueryBuilder() on null

/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/src/Zicht/Bundle/UrlBundle/Aliasing/Listener.php:220
/home/jochem/Code/zicht/krollermuller.nl/vendor/zicht/url-bundle/tests/ZichtTest/Bundle/UrlBundle/Aliasing/ListenerTest.php:213

FAILURES!
Tests: 143, Assertions: 276, Errors: 8.
```

Reply:
Maar die 8 errors komen nu door jouw aanpassingen, in release/2.x gaan ze nl. wel goed. 

Het is wel zo netjes als je de tests die er zijn ook repareert. Sowieso moet dat want anders faalt de build, en kan een PR eigenlijk niet gemerged worden. 

Op zich, zonder nieuwe tests te introduceren, maar door het toevoegen van aanpassingen terwijl de bestaande tests goed blijven gaan heb je ook getest dat oude functionaliteit blijft werken. Dus volgens mij ben je er voor de helft als je de bestaande tests repareert. 

Zie je niet de mogelijkheid om een nieuwe test te introduceren die een scenario test dat met jouw nieuwe code te maken heeft? Hoeft natuurlijk niet in de (wellicht onbegrijpelijke) tests gestopt te worden die er nu zijn.

Reply:
Aha, op PHP 5.6 doet ie het wel... Iemand heeft er doodleuk `php: ^5.6 || ^7` van gemaakt...

Reply:
Je hebt gelijk @erik-zicht, de tests setup moest aangepast worden door de interne aanpassingen in de listener. Heb de tests herschreven en gaat nu goed. Ik heb issue #63 aangemaakt omdat ik denk dat de tests dan beter en nuttiger worden.

Reply:
Jenkins gaat nog steeds flippen want de jobs lopen uit hun geheugen...